### PR TITLE
Adding system metrics support

### LINKED ISSUE #87: System metrics support
**What would you like to be added**:

We should add some metrics for tracking the health of the system, some of them like:
- Rolling update duration
- Recreate group times
- HPA related? Like scaling up/down times & duration?

Please add more if you think so.

**Why is this needed**:

Providing insight of the system

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by Edwinhr716**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes

Fixes #87 

#### Does this PR introduce a user-facing change?

```release-note
adding recreate_group_times, rolling_update_duration metrics, and replica_ready_status_duration
```

### CODE REVIEW

**Comment by ahg-g** [comment]
is this ready for review? asking because it is still marked as WIP

**Comment by Edwinhr716** [comment]
not yet, currently working on adding a metric for latency for pods to be scheduled and become ready, and also plan on adding unit tests

**Comment by Edwinhr716** [comment]
Done with most of it, PTAL. Only missing unit test for `ReadyReplicaStatus` but want to make sure the implementation looks good first

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+	recreateGroupTimes = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Subsystem: "lws",
+			Name:      "recreate_group_times",
+			Help:      "number of times a group has been recreated",
+		}, []string{"leadername"},
+	)
+
+	replicaReadyStatusDuration = prometheus.NewHistogramVec(
```
Comment (liurupeng) [inline pkg/metrics/metrics.go:47]: add a comment here about what is the metrics record, like "the latency for a pod group to be ready after creation"

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	rollingUpdateDuration.WithLabelValues(hash).Observe(duration.Seconds())
+}
+
+func RecreatingGroup(leaderName string) {
+	recreateGroupTimes.WithLabelValues(leaderName).Inc()
+}
+
+func ReplicaReadyStatus(readyPods []corev1.Pod, startTime metav1.Time) {
+	for _, pod := range readyPods {
+		readyTime := podutils.PodReadyConditionLastTransitionTime(pod).Time
```
Comment (liurupeng) [inline pkg/metrics/metrics.go:66]: when the leader pod is ready, its worker statefulset may not be ready right? the latency here calculates the leader pods' latency to be ready

Comment (liurupeng) [reply pkg/metrics/metrics.go:66]: we may need to pass the worker sts and leader pod as a pair and make sure that latency=worker_sts_ready-leader_pod_created, wdyt

File: pkg/controllers/leaderworkerset_controller.go [hunk 3]
```diff
@@ -569,3 +573,12 @@ func templateUpdated(sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerS
 func replicasUpdated(sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet) bool {
 	return *sts.Spec.Replicas != *lws.Spec.Replicas
 }
+
+func getLastTransitionTime(conditionType string, lws *leaderworkerset.LeaderWorkerSet) metav1.Time {
```
Comment (liurupeng) [inline pkg/controllers/leaderworkerset_controller.go:577]: maybe move this to a util file?

File: pkg/controllers/leaderworkerset_controller.go [hunk 4]
```diff
@@ -389,6 +390,9 @@ func (r *LeaderWorkerSetReconciler) updateConditions(ctx context.Context, lws *l
 	updateCondition := setCondition(lws, condition)
 	// if condition changed, record events
 	if updateCondition {
+		if updatedCount == int(*lws.Spec.Replicas) {
+			metrics.ReplicaReadyStatus(readyLeaderPods, getLastTransitionTime(string(leaderworkerset.LeaderWorkerSetProgressing), lws))
```
Comment (liurupeng) [inline pkg/controllers/leaderworkerset_controller.go:394]: the start time here is when the lws just progress, there might be a possibility that the leader pod is not scheduled.

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (liurupeng) [reply pkg/metrics/metrics.go:47]: but overall, I start to think about whether it's necessary to have this metrics for now.

File: pkg/metrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+var (
+	rollingUpdateDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Subsystem: "lws",
+			Name:      "rolling_update_duration",
+			Help:      "Duration of rolling updates",
+		}, []string{"hash"},
+	)
+
+	recreateGroupTimes = prometheus.NewCounterVec(
```
Comment (liurupeng) [inline pkg/metrics/metrics.go:39]: this is not called anywhere else right?

Comment (Edwinhr716) [reply pkg/metrics/metrics.go:39]: Yeah, adding it to register it, but waiting on #106 to implement it.

File: pkg/metrics/metrics.go (hunk 2 above)
Comment (Edwinhr716) [reply pkg/metrics/metrics.go:66]: The LeaderPod is only added into the list that is passed if the worker sts is ready, that is checked [here](https://github.com/kubernetes-sigs/lws/pull/97/files#diff-636763c4aab72861dbe05856e71e11abf551830d79fdcd848f3133ffc9c2d6e1L359) 

I guess I'm a bit confused with what the replica latency measures,  I thought it was from when the lws object was created/edited to when the replica is ready. So is it from when the leader statefulset is created to when the worker sts is ready?

File: pkg/controllers/leaderworkerset_controller.go (hunk 4 above)
Comment (Edwinhr716) [reply pkg/controllers/leaderworkerset_controller.go:394]: Asked a similar question in other comment: what is the replica latency actually measuring? I thought it was from lws creation to when the replica is ready

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (Edwinhr716) [reply pkg/metrics/metrics.go:47]: the help attribute of the object describes what it records, but I can add a comment for it as well

File: pkg/controllers/leaderworkerset_controller.go (hunk 3 above)
Comment (Edwinhr716) [reply pkg/controllers/leaderworkerset_controller.go:577]: Was thinking about that as well, can create an lws utils file similar to the one that exists for pods, and move this function, `templateUpdated` and `replicasUpdated` to it in a different PR.

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (ahg-g) [reply pkg/metrics/metrics.go:47]: I don't think we need such a metric. The time it takes for the pod to become ready is not a characteristic of the lws controller.

**Review by ahg-g — COMMENTED** [review]
I don't know if we really need to add our own metrics at this stage, I would focus on making sure that we are getting the metrics that controller-runtime publishes by default; see https://book.kubebuilder.io/reference/metrics-reference

Those metrics cover a lot of what we need already for monitoring the health of the operator, like how long a reconcile call takes, and I think the first PR should focus on that only and it should be included in the 0.2.0 release.

**Comment by kerthcet** [comment]
+1 to add these metrics later since not part of the 0.2.0.

**Comment by liurupeng** [comment]
/hold
As discussed with Edwin, hold the PR for now, thanks @Edwinhr716

**Comment by liurupeng** [comment]
close for now, @Edwinhr716

**Pull request closed without merging**